File: src/registry/paseo-registry.ts
import { access } from "node:fs/promises";
import path from "node:path";
import { z } from "zod";
import { DirectTcpHostConnectionSchema } from "@getpaseo/protocol/host-connection-schema";
import type { HostEntry } from "../config/host-entry.js";
import { errorText } from "../error-text.js";
import { decodeLocalStorageValue, localStorageKey } from "./local-storage.js";
import { readLevelDbValue } from "./leveldb-reader.js";

/**
 * Reads the Paseo desktop app's host registry out of its Chromium
 * localStorage.
 *
 * This is an unsupported surface: the app persists the registry through
 * async-storage, which on Electron is plain localStorage, and nothing upstream
 * promises the location, the key, or the value encoding. The payload's *shape*
 * is safe, because it is the published protocol schemas; the way we reach it
 * is not. See docs/superpowers/2026-08-19-registry-sync-design.md.
 */

const ORIGIN = "paseo://app";
const REGISTRY_KEY = "@paseo:daemon-registry";

/**
 * The shipped app's support directory. A development build of the desktop
 * package is deliberately not probed: it loads its window from the dev
 * server rather than `paseo://app`, so its localStorage is keyed under that
 * origin and the record this module looks up would never be found there --
 * a probe that succeeds would only turn "not installed" into a misleading
 * "no hosts yet".
 */
const APP_DIR = "Paseo";
const LOCAL_STORAGE_SUBPATH = path.join("Local Storage", "leveldb");

/**
 * The two connection shapes the tray can actually dial, the two it knows it
 * cannot, and a catch-all for whatever the Paseo app adds next.
 *
 * `directTcp` reuses the published `DirectTcpHostConnectionSchema` rather
 * than redefining the shape locally — the same schema `src/config/host-entry.ts`
 * uses for the app's own config — so a protocol change surfaces as a type
 * error here instead of silently drifting. `relay`, `directSocket`, and
 * `directPipe` have no published equivalent in this SDK version, so those
 * stay hand-rolled. `id` is optional on every shape because the app's own
 * on-disk schema makes it optional and regenerates it on load; the tray only
 * needs it to honour `preferredConnectionId`.
 *
 * The catch-all matters because the desktop app is not version-pinned: it
 * writes this database on its own release cadence, and a connection kind the
 * tray has never heard of must reduce to "this host has no connection the
 * menu bar can use", not to a rejected registry that takes every other host
 * down with it. A *known* kind with a malformed shape is still rejected, at
 * the profile level, so the row can say what was wrong.
 */
const KnownConnectionSchema = z.discriminatedUnion("type", [
  DirectTcpHostConnectionSchema.extend({ id: z.string().optional() }),
  z.object({
    id: z.string().optional(),
    type: z.literal("relay"),
    relayEndpoint: z.string(),
    useTls: z.boolean().optional(),
    daemonPublicKeyB64: z.string(),
  }),
  z.object({ id: z.string().optional(), type: z.literal("directSocket"), path: z.string() }),
  z.object({ id: z.string().optional(), type: z.literal("directPipe"), path: z.string() }),
]);
const KNOWN_CONNECTION_TYPES = new Set(["directTcp", "relay", "directSocket", "directPipe"]);
const UnknownConnectionSchema = z.object({
  id: z.string().optional(),
  // `abort` matters: a non-aborting refine failure makes zod report this
  // branch alone as the union's verdict, hiding the known branch's real
  // complaint ("endpoint: required") behind a bare "Invalid input".
  type: z.string().refine((type) => !KNOWN_CONNECTION_TYPES.has(type), {
    message: "known connection types must match their own schema",
    abort: true,
  }),
});
const RegistryConnectionSchema = z.union([KnownConnectionSchema, UnknownConnectionSchema]);

/**
 * One stored profile. `serverId` is trimmed because the app trims it on
 * load, and the tray sends it verbatim as a relay session id; `label` admits
 * `null` because the app's schema does, and an absent, null, or empty label
 * all mean "no name" downstream.
 */
const HostProfileSchema = z.object({
  serverId: z.string().trim().min(1),
  label: z.string().nullable().optional(),
  connections: z.array(RegistryConnectionSchema),
  preferredConnectionId: z.string().nullable().optional(),
});

export interface RegistrySnapshot {
  hosts: HostEntry[];
  /** Hosts the tray cannot dial, phrased for the error row. Never silent. */
  failures: string[];
  /**
   * Set when the hosts above were read out of a database that was partly
   * unreadable, so they may be superseded. The hosts are still applied — a
   * healthy file's answer beats no answer — but the row has to say so, or a
   * host the user deleted in the Paseo app stays in the tray looking healthy.
   */
  warning?: string;
}

type RegistryConnection = z.infer<typeof RegistryConnectionSchema>;
type DialableConnection = Extract<RegistryConnection, { type: "directTcp" | "relay" }>;
type HostProfile = z.infer<typeof HostProfileSchema>;

function isDialable(connection: RegistryConnection): connection is DialableConnection {
  return connection.type === "directTcp" || connection.type === "relay";
}

/**
 * Picks the connection to dial: the profile's preference when the tray
 * supports it, else the first supported one. A profile can name a preference
 * the tray cannot use (a unix socket) while still carrying a usable relay, and
 * dropping that host would lose a working connection for no reason.
 */
function chooseConnection(profile: HostProfile): DialableConnection | null {
  const supported = profile.connections.filter(isDialable);
  const preferred = supported.find((connection) => connection.id === profile.preferredConnectionId);
  return preferred ?? supported[0] ?? null;
}

/**
 * The best name for a profile that may not have parsed: its label if it has
 * one, its serverId if it has one, else its position. Empty strings do not
 * count as names, or the failure row would open with a bare dash.
 */
function describeProfile(candidate: unknown, index: number): string {
  if (typeof candidate === "object" && candidate !== null) {
    const { label, serverId } = candidate as { label?: unknown; serverId?: unknown };
    if (typeof label === "string" && label.trim() !== "") return label;
    if (typeof serverId === "string" && serverId.trim() !== "") return serverId;
  }
  return `profile ${index + 1}`;
}

/**
 * Flattens a profile's issues into "path: message" lines. A connection that
 * matches no branch of the union reports as one `invalid_union` issue whose
 * detail sits a level down, and that detail -- "endpoint: required" -- is
 * the part the row needs.
 */
function describeIssues(issues: z.core.$ZodIssue[]): string[] {
  return issues.flatMap((issue) => {
    if (issue.code === "invalid_union") return issue.errors.flatMap(describeIssues);
    return [`${issue.path.map(String).join(".") || "profile"}: ${issue.message}`];
  });
}

export function hostEntriesFromRegistry(json: string): RegistrySnapshot {
  const parsed: unknown = JSON.parse(json);
  // Anything but an array means the record is not the registry we know how
  // to read at all, so that is a whole-read failure. One profile that does
  // not parse is not: every other host is still exactly as usable as before,
  // and losing them over a sibling is the silent cap this project forbids.
  const candidates = z.array(z.unknown()).parse(parsed);

  const hosts: HostEntry[] = [];
  const failures: string[] = [];
  const seenServerIds = new Set<string>();

  for (const [index, candidate] of candidates.entries()) {
    const name = describeProfile(candidate, index);
    const result = HostProfileSchema.safeParse(candidate);
    if (!result.success) {
      failures.push(`${name} — could not be read (${describeIssues(result.error.issues).join("; ")})`);
      continue;
    }
    const profile = result.data;

    const connection = chooseConnection(profile);
    if (!connection) {
      const kinds = profile.connections.map((entry) => entry.type);
      const has = kinds.length > 0 ? ` (has: ${kinds.join(", ")})` : "";
      failures.push(`${name} — no connection the menu bar can use${has}`);
      continue;
    }

    // Two profiles for one daemon is a shape the Paseo app can hold, and the
    // id keys the fleet's connection map: the second entry would overwrite
    // the first, leaving one live socket wearing the other's label, type, and
    // web base url. The first profile wins and the second is named, because
    // dropping it quietly is the silent cap this project forbids.
    if (seenServerIds.has(profile.serverId)) {
      failures.push(
        `${name} — a second profile for host ${profile.serverId}; the menu bar shows the first`,
      );
      continue;
    }
    seenServerIds.add(profile.serverId);

    // The id is the serverId, never the connection id: distinct hosts share
    // the identical connection id `relay:wss:relay.paseo.sh:443`, which the
    // duplicate-id check rejects outright. serverId is unique per daemon and
    // stable across restarts, which also keeps the derived clientId stable so
    // the daemon resumes sessions instead of starting new ones.
    //
    // `HostEntrySchema` requires a label to be non-empty when present, so an
    // empty or null one is left out rather than carried through.
    const base = { id: profile.serverId, ...(profile.label ? { label: profile.label } : {}) };

    if (connection.type === "directTcp") {
      hosts.push({
        ...base,
        type: "directTcp",
        endpoint: connection.endpoint,
        // `DirectTcpHostConnectionSchema` declares `useTls` with
        // `.default(false)`, so the parsed output is already a plain
        // boolean — no fallback needed here.
        useTls: connection.useTls,
        ...(connection.password !== undefined ? { password: connection.password } : {}),
      });
    } else {
      hosts.push({
        ...base,
        type: "relay",
        offer: {
          v: 2,
          serverId: profile.serverId,
          daemonPublicKeyB64: connection.daemonPublicKeyB64,
          relay: {
            endpoint: connection.relayEndpoint,
            ...(connection.useTls === undefined ? {} : { useTls: connection.useTls }),
          },
        },
      });
    }
  }

  return { hosts, failures };
}

/**
 * The leveldb directory of the installed Paseo app.
 *
 * Throws when it is absent: that is a distinct, actionable state ("the app
 * is not installed") rather than an empty registry, and the tray says so.
 * Any other reason the directory cannot be reached -- a permission change, a
 * file sitting where the directory should be -- is a different problem with
 * a different fix, so it is reported with its own error rather than folded
 * into "not found".
 */
export async function registryLevelDbDir(appSupportDir: string): Promise<string> {
  const dir = path.join(appSupportDir, APP_DIR, LOCAL_STORAGE_SUBPATH);
  try {
    await access(dir);
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === "ENOENT") {
      throw new Error(`Paseo desktop app not found.\n\nLooked in:\n${dir}`);
    }
    throw new Error(`Could not open the Paseo app's storage at ${dir}: ${errorText(error)}`, {
      cause: error,
    });
  }
  return dir;
}

/** `null` means the app is installed but has never stored a registry. */
export async function readRegistry(appSupportDir: string): Promise<RegistrySnapshot | null> {
  const dir = await registryLevelDbDir(appSupportDir);
  const read = await readLevelDbValue(dir, localStorageKey(ORIGIN, REGISTRY_KEY));
  if (read.value === null) return null;
  const snapshot = hostEntriesFromRegistry(decodeLocalStorageValue(read.value));
  // A value found next to an unreadable file is usable but not trustworthy as
  // the last word; the caller applies it and shows the detail.
  return read.parseFailure === null ? snapshot : { ...snapshot, warning: read.parseFailure };
}

File: PaseoIconPackage/Sources/PaseoIconCore/Registry/PaseoRegistry.swift
import Foundation

/// The hosts the Paseo desktop app has stored, as the tray can use them.
public struct RegistrySnapshot: Equatable, Sendable {
    public var hosts: [HostEntry]
    /// Hosts the tray cannot dial, phrased for the error row. Never silent.
    public var failures: [String]
    /// Set when the hosts above were read out of a database that was partly
    /// unreadable, so they may be superseded.
    public var warning: String?

    public init(hosts: [HostEntry], failures: [String], warning: String? = nil) {
        self.hosts = hosts
        self.failures = failures
        self.warning = warning
    }
}

public enum RegistryError: MessageError, Equatable {
    case notAnArray
    case invalidJSON(String)
    case appNotFound(dir: String)
    case storageUnreachable(dir: String, detail: String)

    public var message: String {
        switch self {
        case .notAnArray: "The registry record is not an array of profiles"
        case .invalidJSON(let detail): "The registry record is not valid JSON: \(detail)"
        case .appNotFound(let dir): "Paseo desktop app not found.\n\nLooked in:\n\(dir)"
        case .storageUnreachable(let dir, let detail): "Could not open the Paseo app's storage at \(dir): \(detail)"
        }
    }
}

/// Reads the Paseo desktop app's host registry out of its Chromium
/// localStorage. An unsupported surface: nothing upstream promises the
/// location, the key, or the value encoding. The payload's shape is the
/// published connection schemas, and that part is safe.
public enum PaseoRegistry {
    public static let origin = "paseo://app"
    public static let registryKey = "@paseo:daemon-registry"
    /// The shipped app's support directory. A development build loads from
    /// the dev server and keys its storage under another origin, so it is
    /// deliberately not probed.
    static let appDirectory = "Paseo"
    static let localStorageSubpath = "Local Storage/leveldb"
    static let knownConnectionTypes: Set<String> = ["directTcp", "relay", "directSocket", "directPipe"]

    /// The leveldb directory of the installed Paseo app. Absent is a distinct,
    /// actionable state ("the app is not installed"); any other reason it
    /// cannot be reached is reported with its own error.
    public static func levelDbDir(appSupportDir: String) throws -> String {
        let dir = (appSupportDir as NSString).appendingPathComponent(appDirectory) + "/" + localStorageSubpath
        if access(dir, F_OK) == 0 { return dir }
        let code = errno
        if code == ENOENT { throw RegistryError.appNotFound(dir: dir) }
        let detail = "\(String(cString: strerror(code))) (\(posixName(code)))"
        throw RegistryError.storageUnreachable(dir: dir, detail: detail)
    }

    /// Nil means the app is installed but has never stored a registry.
    public static func read(appSupportDir: String, fileSystem: any FileSystem = LocalFileSystem()) throws -> RegistrySnapshot? {
        let dir = try levelDbDir(appSupportDir: appSupportDir)
        let result = try LevelDBReader.readValue(directory: dir, userKey: LocalStorage.key(origin: origin, key: registryKey), fileSystem: fileSystem)
        guard let value = result.value else { return nil }
        var snapshot = try hostEntries(fromJSON: try LocalStorage.decodeValue(value))
        // A value found next to an unreadable file is usable but not the last word.
        snapshot.warning = result.parseFailure
        return snapshot
    }

    /// Maps the stored profiles to host entries. Anything but an array fails
    /// the whole read; one profile that does not parse is named and skipped,
    /// because losing the others over a sibling is the silent cap this
    /// project forbids.
    public static func hostEntries(fromJSON json: String) throws -> RegistrySnapshot {
        let parsed: Any
        do {
            parsed = try JSONSerialization.jsonObject(with: Data(json.utf8), options: [.fragmentsAllowed])
        } catch {
            throw RegistryError.invalidJSON(errorText(error))
        }
        guard let candidates = parsed as? [Any] else { throw RegistryError.notAnArray }

        var hosts: [HostEntry] = []
        var failures: [String] = []
        var seenServerIds = Set<String>()

        for (index, candidate) in candidates.enumerated() {
            let name = describeProfile(candidate, index: index)
            let profile: Profile
            do {
                profile = try Profile(candidate)
            } catch let error as ProfileError {
                failures.append("\(name) — could not be read (\(error.issues.joined(separator: "; ")))")
                continue
            }

            guard let connection = profile.chooseConnection() else {
                let kinds = profile.connections.map(\.type)
                let has = kinds.isEmpty ? "" : " (has: \(kinds.joined(separator: ", ")))"
                failures.append("\(name) — no connection the menu bar can use\(has)")
                continue
            }

            // Two profiles for one daemon: the id keys the fleet's connection
            // map, so the first wins and the second is named.
            if seenServerIds.contains(profile.serverId) {
                failures.append("\(name) — a second profile for host \(profile.serverId); the menu bar shows the first")
                continue
            }
            seenServerIds.insert(profile.serverId)

            // The id is the serverId, never the connection id: distinct hosts
            // share the identical relay connection id. An empty label is left
            // out rather than carried through.
            let label = profile.label.flatMap { $0.isEmpty ? nil : $0 }
            switch connection {
            case .directTcp(let endpoint, let useTls, let password):
                hosts.append(.directTcp(id: profile.serverId, label: label, endpoint: endpoint, useTls: useTls, password: password))
            case .relay(let endpoint, let useTls, let daemonPublicKeyB64):
                let offer = ConnectionOffer(
                    serverId: profile.serverId,
                    daemonPublicKeyB64: daemonPublicKeyB64,
                    relay: .init(endpoint: endpoint, useTls: useTls)
                )
                // Validated here, per profile, rather than only in
                // `AppConfig.validate`. That one throws for the whole list, and
                // `RegistrySession.readOnce` catches it and returns before
                // applying anything — so one relay profile whose pairing never
                // finished cost every other host, including the local daemon.
                // "Not to zero hosts" has to hold at this level too, and the
                // guard in `AppConfig` then becomes the unreachable last line
                // of defence it reads as.
                do {
                    _ = try offer.validated()
                } catch let error as ConnectionOfferError {
                    failures.append("\(name) — \(error.message)")
                    continue
                }
                hosts.append(.relay(id: profile.serverId, label: label, offer: offer))
            }
        }
        return RegistrySnapshot(hosts: hosts, failures: failures)
    }

    // MARK: - Profile parsing

    private struct ProfileError: Error {
        let issues: [String]
    }

    private enum Dialable {
        case directTcp(endpoint: String, useTls: Bool, password: String?)
        case relay(endpoint: String, useTls: Bool?, daemonPublicKeyB64: String)
    }

    private struct Connection {
        let id: String?
        let type: String
        let dialable: Dialable?
    }

    /// One stored profile, validated the way the published schemas validated
    /// it: a known connection kind with a malformed shape is an issue, an
    /// unknown kind is merely unusable.
    private struct Profile {
        let serverId: String
        let label: String?
        let connections: [Connection]
        let preferredConnectionId: String?

        init(_ candidate: Any) throws {
            guard let object = candidate as? [String: Any] else { throw ProfileError(issues: ["profile: expected an object"]) }
            var issues: [String] = []

            var serverId = ""
            if let raw = object["serverId"] as? String {
                serverId = raw.trimmingCharacters(in: .whitespacesAndNewlines)
                if serverId.isEmpty { issues.append("serverId: must not be empty") }
            } else {
                issues.append("serverId: required")
            }

            var label: String?
            if let raw = object["label"] {
                if let text = raw as? String { label = text } else if !(raw is NSNull) { issues.append("label: expected a string") }
            }

            var connections: [Connection] = []
            if let rawConnections = object["connections"] as? [Any] {
                for (index, raw) in rawConnections.enumerated() {
                    guard let entry = raw as? [String: Any], let type = entry["type"] as? String else {
                        issues.append("connections.\(index): expected an object with a type")
                        continue
                    }
                    let id = entry["id"] as? String
                    let prefix = "connections.\(index)"
                    switch type {
                    case "directTcp":
                        guard let endpoint = entry["endpoint"] as? String else { issues.append("\(prefix).endpoint: required"); continue }
                        var useTls = false
                        if let raw = entry["useTls"] { if let flag = PaseoRegistry.boolValue(raw) { useTls = flag } else { issues.append("\(prefix).useTls: expected a boolean"); continue } }
                        var password: String?
                        if let raw = entry["password"] { if let text = raw as? String { password = text } else { issues.append("\(prefix).password: expected a string"); continue } }
                        connections.append(Connection(id: id, type: type, dialable: .directTcp(endpoint: endpoint, useTls: useTls, password: password)))
                    case "relay":
                        guard let endpoint = entry["relayEndpoint"] as? String else { issues.append("\(prefix).relayEndpoint: required"); continue }
                        guard let key = entry["daemonPublicKeyB64"] as? String else { issues.append("\(prefix).daemonPublicKeyB64: required"); continue }
                        var useTls: Bool?
                        if let raw = entry["useTls"] { if let flag = PaseoRegistry.boolValue(raw) { useTls = flag } else { issues.append("\(prefix).useTls: expected a boolean"); continue } }
                        connections.append(Connection(id: id, type: type, dialable: .relay(endpoint: endpoint, useTls: useTls, daemonPublicKeyB64: key)))
                    case "directSocket", "directPipe":
                        guard entry["path"] is String else { issues.append("\(prefix).path: required"); continue }
                        connections.append(Connection(id: id, type: type, dialable: nil))
                    default:
                        // A kind the tray has never seen reduces to "unusable", not to a rejected profile.
                        connections.append(Connection(id: id, type: type, dialable: nil))
                    }
                }
            } else {
                issues.append("connections: required")
            }

            var preferred: String?
            if let raw = object["preferredConnectionId"] {
                if let text = raw as? String { preferred = text } else if !(raw is NSNull) { issues.append("preferredConnectionId: expected a string") }
            }

            guard issues.isEmpty else { throw ProfileError(issues: issues) }
            self.serverId = serverId
            self.label = label
            self.connections = connections
            self.preferredConnectionId = preferred
        }

        /// The profile's preference when the tray supports it, else the first
        /// supported one.
        func chooseConnection() -> Dialable? {
            let supported = connections.filter { $0.dialable != nil }
            if let preferred = supported.first(where: { $0.id != nil && $0.id == preferredConnectionId }) { return preferred.dialable }
            return supported.first?.dialable
        }
    }

    /// The best name for a profile that may not have parsed: label, serverId,
    /// else its position.
    private static func describeProfile(_ candidate: Any, index: Int) -> String {
        if let object = candidate as? [String: Any] {
            if let label = object["label"] as? String, !label.trimmingCharacters(in: .whitespaces).isEmpty { return label }
            if let serverId = object["serverId"] as? String, !serverId.trimmingCharacters(in: .whitespaces).isEmpty { return serverId }
        }
        return "profile \(index + 1)"
    }

    /// `JSONSerialization` returns `NSNumber` for both booleans and integers,
    /// and `as? Bool` bridges 0 and 1 as happily as false and true — so a
    /// profile written with `"useTls": 1` would be accepted where the
    /// published schema rejects it. `CFBoolean` is the only thing that is
    /// really a boolean.
    private static func boolValue(_ raw: Any) -> Bool? {
        guard let number = raw as? NSNumber, CFGetTypeID(number) == CFBooleanGetTypeID() else { return nil }
        return number.boolValue
    }

    private static func posixName(_ code: Int32) -> String {
        switch code {
        case ENOTDIR: "ENOTDIR"
        case EACCES: "EACCES"
        case ELOOP: "ELOOP"
        default: "errno \(code)"
        }
    }
}
